### PR TITLE
Feature/date index iterator

### PR DESCRIPTION
**Pull request by apmoriarty**

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/util/DateIndexIterator.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+/**
+ * An iterator optimized for the {@link TableName#DATE_INDEX}
+ * <p>
+ * Both {@link DateIndexHelper#getTypeDescription(String, Date, Date, Set)} and
+ * {@link DateIndexHelper#getShardsAndDaysHint(String, Date, Date, Date, Date, Set)} support a datatype filter.
+ * <p>
+ * Filtering based on the start/end date is done via the {@link #MINIMUM_DATE} amd {@link #MAXIMUM_DATE} options.
+ * <p>
+ * Filtering based on the field is done via the {@link #FIELD} option.
+ */
```
Comment (lbschanno) [inline warehouse/query-core/src/main/java/datawave/query/util/DateIndexIterator.java:35]: Would like to see the javadoc include information on how the `time.travel.enabled` option impacts the iterator's functionality.

File: warehouse/query-core/src/main/java/datawave/query/util/DateTypeDescriptionIterator.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    private DateTypeDescription description = new DateTypeDescription();
+
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+        this.source = source;
+
+        if (options.containsKey(DATATYPE_FILTER)) {
+            String option = options.get(DATATYPE_FILTER);
+            datatypes = new HashSet<>(Splitter.on(',').splitToList(option));
```
Comment (lbschanno) [inline warehouse/query-core/src/main/java/datawave/query/util/DateTypeDescriptionIterator.java]: Splitter instances are thread-safe and immutable, so you could have a static final constant of a Splitter for reuse.

File: warehouse/query-core/src/main/java/datawave/query/util/DateIndexHelper.java [hunk 3]
```diff
@@ -170,39 +173,93 @@ public DateIndexHelper initialize(AccumuloClient client, String dateIndexTableNa
...
-            builder.append(", dateRange").append(Arrays.asList(dateRange));
-            return builder.toString();
+            return "field: " + fields + ", dateRange[" + start + "," + end + "]";
+        }
+
+        public String serializeToString() {
+            StringBuilder sb = new StringBuilder();
+            sb.append(start).append(',');
+            sb.append(end).append(',');
+            sb.append(Joiner.on(',').join(fields));
```
Comment (lbschanno) [inline warehouse/query-core/src/main/java/datawave/query/util/DateIndexHelper.java]: Joiner instances are thread-safe and immutable, so you could have a static final constant for reuse throughout the class.

File: warehouse/query-core/src/test/java/datawave/query/util/DateIndexHelperTest.java [hunk 4]
```diff
@@ -225,4 +304,53 @@ public void testDateIndexHelperHint() throws Exception {
                         DateIndexUtil.getBeginDate("20090101"), DateIndexUtil.getEndDate("20120101"), Collections.singleton("test"));
         Assert.assertEquals("", hint);
     }
+
+    @Test
+    public void testDateIndexHelperHintWithIterator() throws Exception {
+        DateIndexHelper helper = this.dateIndexHelperFactory.createDateIndexHelper().initialize(client, TableName.DATE_INDEX, Collections.singleton(auths), 2,
+                        0.9f);
+        helper.setUseIterator(true);
+
```
Comment (lbschanno) [inline warehouse/query-core/src/test/java/datawave/query/util/DateIndexHelperTest.java:315]: Can you add some documentation about what scenarios/edge cases each of these different asserts are verifying?

File: warehouse/query-core/src/test/java/datawave/query/util/DateIndexIteratorTest.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+ * Tests for the {@link DateIndexIterator}
+ */
+public class DateIndexIteratorTest {
+
+    private static final Logger log = LoggerFactory.getLogger(DateIndexIteratorTest.class);
+
+    private final SortedMap<Key,Value> data = new TreeMap<>();
+    private final Map<String,String> options = new HashMap<>();
+
+    private SortedKeyValueIterator<Key,Value> source;
```
Comment (lbschanno) [inline warehouse/query-core/src/test/java/datawave/query/util/DateIndexIteratorTest.java]: This is never used and can be deleted.

File: warehouse/query-core/src/test/java/datawave/query/util/DateIndexIteratorTest.java [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        drive("20250404", "LOADED");
+    }
+
+    @Test
+    public void testMultipleLoadedDates() throws IOException {
+        write("20250404", "LOADED", "20250402", "datatype-a", "FIELD_A", 1, 2, 3);
+        write("20250404", "LOADED", "20250403", "datatype-a", "FIELD_A", 4, 5, 6);
+        write("20250404", "LOADED", "20250404", "datatype-a", "FIELD_A", 7, 8, 9);
+        withExpectedNextSeekCounts(4, 1);
+        drive("20250404", "LOADED");
```
Comment (lbschanno) [inline warehouse/query-core/src/test/java/datawave/query/util/DateIndexIteratorTest.java]: `drive()` is a bit ambiguous as a method name. Can this be renamed to `assertExpected()` or something similar?

File: warehouse/query-core/src/test/java/datawave/query/util/DateIndexIteratorTest.java [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+        Collection<ByteSequence> columnFamily = Collections.singleton(new ArrayByteSequence(type.getBytes()));
+
+        StatDateIndexIterator iter = new StatDateIndexIterator();
+        iter.init(source, options, env);
+        iter.seek(range, columnFamily, true);
+
+        while (iter.hasTop()) {
+            Key key = iter.getTopKey();
+            Value value = iter.getTopValue();
```
Comment (lbschanno) [inline warehouse/query-core/src/test/java/datawave/query/util/DateIndexIteratorTest.java]: The variable `value` is never used. Should this be logged below alongside the key?

File: warehouse/query-core/src/test/java/datawave/query/util/DateIndexIteratorTest.java [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    private void withStartEndDate(String startDate, String endDate) {
+        options.put(DateIndexIterator.MINIMUM_DATE, startDate);
+        options.put(DateIndexIterator.MAXIMUM_DATE, endDate);
+    }
+
+    private void withField(String field) {
+        options.put(DateIndexIterator.FIELD, field);
+    }
+
+    private void withTimeTravel(boolean flag) {
```
Comment (lbschanno) [inline warehouse/query-core/src/test/java/datawave/query/util/DateIndexIteratorTest.java:223]: I noticed that this function is never called. Will you be adding tests cases later that test the time travel aspect of the iterator?

File: warehouse/query-core/src/test/java/datawave/query/util/DateTypeDescriptionIteratorTest.java [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+import datawave.query.iterator.SourceManagerTest;
+
+public class DateTypeDescriptionIteratorTest {
+
+    private static final Logger log = LoggerFactory.getLogger(DateTypeDescriptionIteratorTest.class);
+
+    private final SortedMap<Key,Value> data = new TreeMap<>();
+    private final Map<String,String> options = new HashMap<>();
+
+    private SortedKeyValueIterator<Key,Value> source;
```
Comment (lbschanno) [inline warehouse/query-core/src/test/java/datawave/query/util/DateTypeDescriptionIteratorTest.java]: `source` is never used and can be deleted.

File: warehouse/query-core/src/test/java/datawave/query/util/DateTypeDescriptionIteratorTest.java [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+        Key key = new Key(shard, type, date + "\u0000" + datatype + "\u0000" + field);
+        BitSet bitset = new BitSet();
+        for (int offset : offsets) {
+            bitset.set(offset);
+        }
+        Value value = new Value(bitset.toByteArray());
+        data.put(key, value);
+    }
+
+    private class StatDateTypeDescriptionIterator extends DateTypeDescriptionIterator {
```
Comment (lbschanno) [inline warehouse/query-core/src/test/java/datawave/query/util/DateTypeDescriptionIteratorTest.java:172]: This iterator is functionally identical to `DateIndexIteratorTest.StatDateIndexIterator`. I think it's a candidate to be extracted to a top-level class that can then be used by different test classes.

File: warehouse/query-core/src/main/java/datawave/query/util/DateIndexHelper.java [hunk 11]
```diff
@@ -170,39 +173,93 @@ public DateIndexHelper initialize(AccumuloClient client, String dateIndexTableNa
...
+
+        /**
+         * Method used to set the default values for the start and end date if none were found
+         *
+         * @param startDate
+         *            the start date
+         * @param endDate
+         *            the end date
+         */
+        public void setStartEndDate(Date startDate, Date endDate) {
```
Comment (lbschanno) [inline warehouse/query-core/src/main/java/datawave/query/util/DateIndexHelper.java]: I did not see a check in `DateIndexHelper.getDateTypeDescription()` to assert that the arguments `begin` and `end` are not null. I suggest adding checks here to only update start/end if startDate/endDate are not null to avoid NPEs when formatting the dates. Additionally, the method name could make someone think they are always setting the start and end date if they don't read the documentation. Perhaps a name like `replaceNullDateRangeEndpoints()` or something like that?

**Review by lbschanno — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/util/DateIndexHelper.java [hunk 12]
```diff
@@ -170,39 +173,93 @@ public DateIndexHelper initialize(AccumuloClient client, String dateIndexTableNa
...
+            }
+        }
+
+        /**
+         * Method that updates the start and end date if the provided date is beyond either bound
+         *
+         * @param date
+         *            the date
+         */
+        public void updateStartEndDate(String date) {
```
Comment (lbschanno) [inline warehouse/query-core/src/main/java/datawave/query/util/DateIndexHelper.java]: This method name is a bit ambiguous for what it does. Perhaps `ensureDateRangeIncludes(String date)`?

**Review by lbschanno — CHANGES_REQUESTED** [review]

**Review by lbschanno — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/util/DateIndexIterator.java [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+                    source.next();
+                    continue;
+                }
+            }
+
+            source.next();
+        }
+    }
+
+    private String getRowSansShard(Key key) {
```
Comment (jalphonso) [inline warehouse/query-core/src/main/java/datawave/query/util/DateIndexIterator.java]: should this be `getRowScansShard`?

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/util/DateIndexIterator.java]: No, this is the row without the shard, hence the 'sans'. I will rename this for clarity.

Comment (jalphonso) [reply warehouse/query-core/src/main/java/datawave/query/util/DateIndexIterator.java]: oh, lol. thank you for renaming. I wasn't thinking about the latin meaning.

**Review by jalphonso — APPROVED** [review]

**Review by lbschanno — APPROVED** [review]
